File: deepeval/tracing/internal.py
import inspect
from typing import List, Optional


def observe_methods(
    cls,
    span_type: Optional[str] = None,
    allowed_methods: Optional[List[str]] = None,
):
    from deepeval.tracing.tracing import observe

    is_traceable = lambda v: inspect.isfunction(
        v
    ) or inspect.iscoroutinefunction(v)

    methods = {
        k: v
        for k, v in cls.__dict__.items()
        if not k.startswith("__") and is_traceable(v)
    }

    if allowed_methods is not None:
        methods = {k: v for k, v in methods.items() if k in allowed_methods}

    for name, method in methods.items():
        if getattr(method, "_is_deepeval_observed", False):
            continue
        setattr(
            cls,
            name,
            observe(
                type=span_type,
                _drop_if_root=True,
                _internal=True,
            )(method),
        )

File: typescript/src/tracing/internal.ts
// Opt-in tracing of DeepEval's own metric and model methods, the counterpart of
// Python's `deepeval/tracing/internal.py`.

import { observe, type SpanType } from "@/tracing/tracing";
import { isTraceInternalEnabled } from "@/tracing/utils";

const instrumented = new WeakSet<object>();

/**
 * Wrap the named methods of `instance`'s own class in internal spans. Called
 * from a base-class constructor, this reaches the concrete subclass prototype.
 *
 * Only safe for methods that already return a promise, since an observed method
 * always does.
 */
export function observeMethods(
  instance: object,
  options: { spanType?: SpanType | string; methods: string[] },
): void {
  if (!isTraceInternalEnabled()) return;

  const prototype = Object.getPrototypeOf(instance);
  if (!prototype || prototype === Object.prototype) return;
  if (instrumented.has(prototype)) return;
  instrumented.add(prototype);

  const className = prototype.constructor?.name ?? "anonymous";

  for (const method of options.methods) {
    const descriptor = Object.getOwnPropertyDescriptor(prototype, method);
    if (!descriptor || typeof descriptor.value !== "function") continue;

    const original = descriptor.value as (...args: any[]) => any;
    Object.defineProperty(prototype, method, {
      ...descriptor,
      value: function (this: unknown, ...args: any[]) {
        return observe({
          type: options.spanType,
          name: `${className}.${method}`,
          _internal: true,
          _dropIfRoot: true,
          fn: (...inner: any[]) => original.apply(this, inner),
        })(...args);
      },
    });
  }
}
